File: Day28/MinDepthOfBinTree.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) { this.val = val; }
}

class MinDepthOfBinTree {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null) return 1 + minDepth(root.right);
        if (root.right == null) return 1 + minDepth(root.left);
        return 1 + Math.min(minDepth(root.left), minDepth(root.right));
    }

    // Build tree from level order input (null represented as "null")
    public static TreeNode buildTree(String[] nodes) {
        if (nodes.length == 0 || nodes[0].equals("null")) return null;
        TreeNode root = new TreeNode(Integer.parseInt(nodes[0]));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;
        while (!queue.isEmpty() && i < nodes.length) {
            TreeNode curr = queue.poll();
            if (i < nodes.length && !nodes[i].equals("null")) {
                curr.left = new TreeNode(Integer.parseInt(nodes[i]));
                queue.add(curr.left);
            }
            i++;
            if (i < nodes.length && !nodes[i].equals("null")) {
                curr.right = new TreeNode(Integer.parseInt(nodes[i]));
                queue.add(curr.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter tree nodes in level order (space separated, 'null' for empty):");
        String[] nodes = sc.nextLine().split(" ");
        TreeNode root = buildTree(nodes);
        MinDepthOfBinTree sol = new MinDepthOfBinTree();
        System.out.println("Minimum Depth of Binary Tree: " + sol.minDepth(root));
        sc.close();
    }
}

File: Day28/MinDepthOfBinTree.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def minDepth(self, root):
        if not root:
            return 0
        if not root.left:
            return 1 + self.minDepth(root.right)
        if not root.right:
            return 1 + self.minDepth(root.left)
        return 1 + min(self.minDepth(root.left), self.minDepth(root.right))

def build_tree(nodes):
    if not nodes or nodes[0] == 'null':
        return None
    root = TreeNode(int(nodes[0]))
    queue = [root]
    i = 1
    while queue and i < len(nodes):
        curr = queue.pop(0)
        if nodes[i] != 'null':
            curr.left = TreeNode(int(nodes[i]))
            queue.append(curr.left)
        i += 1
        if i < len(nodes) and nodes[i] != 'null':
            curr.right = TreeNode(int(nodes[i]))
            queue.append(curr.right)
        i += 1
    return root

if __name__ == "__main__":
    node_str = input("Enter tree nodes in level order (use 'null' for missing):\n")
    nodes = node_str.strip().split()
    root = build_tree(nodes)
    sol = Solution()
    print("Minimum Depth of Binary Tree:", sol.minDepth(root))
